VRF V2 Direct funding docs (#854)
smartcontractkit/documentation: docs/Using Randomness/v2-subscription/design.md

---
layout: nodes.liquid
section: ethereum
date: Last Modified
title: 'Subscription Method'
permalink: 'docs/vrf/v2/subscription/'
whatsnext:
  {
    'Get a Random Number': '/docs/vrf/v2/subscription/get-a-random-number/',
    'Configuration': '/docs/vrf/v2/subscription/configuration/',
  }
metadata:

>   title: 'Generate Random Numbers for Smart Contracts using Chainlink VRF v2 - Subscription method'

Reviewer: ```suggestion
  title: 'Generate Random Numbers for Smart Contracts using Chainlink VRF v2 - Subscription Method'
```

- **Verification gas:** The amount of gas used to verify randomness on-chain.

The gas price depends on current network conditions. The callback gas depends on your callback function, and the number of random values in your request. The cost of each request is final only after the transaction is complete, but you define the limits you are willing to spend for the request with the following variables:

- **Gas lane:** The maximum gas price you are willing to pay for a request in wei. Define this limit by specifying the appropriate `keyHash` in your request. The limits of each gas lane are important for handling gas price spikes when Chainlink VRF bumps the gas price to fulfill your request quickly.

- **Callback gas limit:** Specifies the maximum amount of gas you are willing to spend on the callback request. Define this limit by specifying the `callbackGasLimit` value in your request.

### Request and Receive Data

#### End To End Diagram

![Vrf v2 subscription method end to end diagram](/images/vrf/v2-subscription-e2e.png)

Two types of accounts exist in the Ethereum ecosystem:

> - EOA (Externally Owned Account): an externally owned account has a private key and can control a smart contract. Transactions can only be intiated by EOAs.
> - Smart contract: a smart contract does not have a private key and executes what it has been designed for as a decentralized application.

Reviewer: ```suggestion
- EOA (Externally Owned Account): An externally owned account that has a private key and can control a smart contract. Transactions can only be initiated by the EOA.
- Smart contract: A contract that does not have a private key and executes what it has been designed for as a decentralized application.
```

> The Chainlink VRF v2 solution utilizes both an off-chain and on-chain component:

Reviewer: ```suggestion
The Chainlink VRF v2 solution uses both off-chain and on-chain components:
```

> - [VRF Coordinator (on-chain component)](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/VRFCoordinatorV2.sol): designed to interact with the VRF node. It emits an event when a request for randomness is made then verifies the random number and proof on how it was generated by the VRF Node.
> - VRF Node (off-chain component): Listens for requests by subscribing to the VRF Coordinator event logs and calculates a random number based on the block hash and nonce. The node then sends a transaction to the VRFCoordinator including the random number and a proof on how it was generated.

Reviewer: ```suggestion
- [VRF v2 Coordinator (on-chain component)](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/VRFCoordinatorV2.sol): A contract designed to interact with the VRF node. It emits an event when a request for randomness is made, and then verifies the random number and proof of how it was generated by the VRF Node.
- VRF Node (off-chain component): Listens for requests by subscribing to the VRF Coordinator event logs and calculates a random number based on the block hash and nonce. The node then sends a transaction to the `VRFCoordinator` including the random number and a proof of how it was generated.
```

#### Explanation

Requests to Chainlink VRF v2 follow the [Request & Receive Data](/docs/request-and-receive-data/) cycle. The VRF coordinator processes the request and determines the final charge to your subscription using the following steps:

> 1. The consuming contract must inherit [VRFConsumerBaseV2](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/VRFConsumerBaseV2.sol) and implement the `fulfillRandomWords` function, which is the _callback VRF function_. Submit your VRF request by calling `requestRandomWords` of the [VRF Coordinator](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/VRFCoordinatorV2.sol) with:

Reviewer: This should render properly with all the ordered items as `1.` and two tabs for all child elements.

- `keyHash`: Identifier that maps to a job and a private key on the VRF node and that represents a specified gas lane. If your request is urgent, specify a gas lane with a higher gas price limit. The configuration for your network can be found [here](/docs/vrf/v2/subscription/configuration/#configurations).

> - `s_subscriptionId`: The subscription id that the consuming contract is registered to. LINK funds are deducted from this subscription.

Reviewer: ```suggestion
- `s_subscriptionId`: The subscription ID that the consuming contract is registered to. LINK funds are deducted from this subscription.
```
